### PR TITLE
Port the new hijack from TG, created by citadel. 

### PR DESCRIPTION
**Pull request by Sinestia**
## About The Pull Request
PORTS
https://github.com/tgstation/tgstation/pull/54146
https://github.com/tgstation/tgstation/pull/54435
https://github.com/tgstation/tgstation/pull/54827
https://github.com/tgstation/tgstation/pull/54913
https://github.com/tgstation/tgstation/pull/55214

The ORIGINAL Pr this is all from is here 
https://github.com/Citadel-Station-13/Citadel-Station-13/pull/10649
In a bullet list:

- Old hijack is turned into an "elimination" objective, that is only a custom objective for highlanders
- In order to hijack now, you have to get access to the shuttle console and alt-click it multiple times, while fighting off interruptions. The crew is made aware of this.
- With every bit of progress, you get some extra time given to you. 
- Emergency shuttle console added to BYOS to prevent it from being OP. 
- In a separate PR I just submitted, the shuttle console is made indestructible. I just thought I'd mention it here. 

Some personal notes:
The way its done on an interface level is fairly clunky, imo. And the objective text is downright ugly. The latter could easily be changed here. Former would be a bit more work and design, it was proposed in the discord to give a disk for it, but I don't like the idea of antags having to carry around more items, and it'd have to apply to *all* hijackers. Maybe at minimum it could eventually be added to the actual shuttle console interface.

Works on my testing at least. 

## Why It's Good For The Game
Hijack is pretty flawed.

 It's genuinely often impossible to succeed if you decide to actually do your objective. You've got engineers hiding in the goddamn engine, pAIs, etc. ruining it and making it extremely inconsistent with how its requirement works. 

 This makes it about securing the cockpit for a period of time.

 You might worry about losing the right to murderbone but that's up to the admins, it will probably remain in anyways since it's the easiest way to accomplish this. Maybe stuff like roundstart plasmaflooding would go, but who would honestly miss that?

## Changelog
:cl: silicons, /tg/ coders, port by froststahr
add: New hijack! To do it, you must alt-click the shuttle console a series of times. After that has been successfully done 5 times, the hijack is complete. 
add: After each hack, the shuttle in-transit is delayed slightly, and crew are also notified of what is going on. 
add: Who can hijack? Blood brothers, traitors, changelings, wizards, highlanders, incursions, nukies, heretics
del: Old method of hijack no longer works for ordinary antags, if you want your greentext, you MUST secure the console. Not that that'll be difficult compared to killing everything. 
/:cl:

### CODE REVIEW

**Comment by ghost** [comment]
I want to murderbone :(

**Comment by Effemu** [comment]
hijack is not about greentext its about being free to use and do shit you're not allowed to normally

**Comment by Sinestia** [comment]
> 
> 
> I want to murderbone :(



> 
> 
> hijack is not about greentext its about being free to use and do shit you're not allowed to normally

This does nothing to stop those
The only thing that would would be rule changes

/tg/ antags still do whatever they want 4 example

**Comment by ghost** [comment]
Did you people even read the entire PR description?

**Comment by WhyIsCaeciliusTaken** [comment]
No rule changes would be needed in my opinion if this objective change was merged. You'd still be able to go ham on the crew since you'd have to fight them off at the shuttle still.

**Comment by bloons3** [comment]
> hijack is not about greentext its about being free to use and do shit you're not allowed to normally

This doesn't change people having Hijack objectives.

It makes a new mechanic to hijack the shuttle, which makes it harder to fail the Hijack objective because a single person was hiding on the shuttle.

**Comment by PowerfulBacon** [comment]
Wishgranter should keep the elimination objective, since its the whole backstory behind them. They do not care about stealing a shuttle, they care about killing. As for the rest I guess its cool, you can still emag it then hijack it right?

**Comment by bloons3** [comment]
This is one of the reasons why this was coded
image

**Comment by Sinestia** [comment]
> Wishgranter should keep the elimination objective, since its the whole backstory behind them. They do not care about stealing a shuttle, they care about killing. As for the rest I guess its cool, you can still emag it then hijack it right?

Done

also yeah, emagging still works and is highly desirable to minimize the amount of badmen trying to stop you

**Comment by bloons3** [comment]
Does this still allow antags without emags to hijack or is that a requirement?

**Comment by Sinestia** [comment]
Gotta expel some junk I noticed, I used cherrypick and did it sloppily leaving some useless code.

> Does this still allow antags without emags to hijack or is that a requirement?

No need for an emag to hijack in the first place, your antag datum just has to have the hijack speed variable set to a non-0.

 All the emag does is the same as always, instantly launching the shuttle. 

 There is a quirk with it which is that you don't need the hijack objective to do it. You just have to be one of the antags listed in the CL. Could be changed but there probably isn't any harm leaving it as is. Might get in the way of alliances too.

**Comment by bloons3** [comment]
I would say to leave it, because there might be the possibility of traitors without a hijack objective working with a traitor or other antagonist that does

**Comment by Sinestia** [comment]
Is basically ready to be looked at/tested now i think.

**Comment by ghost** [comment]
if you use A emag on the escape shuttle console could you still have enough time to go through hijack protocol thingy?

**Comment by Sinestia** [comment]
> if you use A emag on the escape shuttle console could you still have enough time to go through hijack protocol thingy?

Yep, emag doesn't make the flight any shorter it just launches it. 

 Time isn't much of an obstacle here, it doesn't take exceptionally long to do a hack, and you get extra time after completing one.

File: code/__DEFINES/antagonists.dm [hunk 1]
```diff
@@ -52,6 +52,14 @@
...
 #define INITIAL_DOM_ATTEMPTS	3
 
+//Shuttle elimination hijacking
+/// Does not stop elimination hijacking but itself won't elimination hijack
+#define ELIMINATION_NEUTRAL 0
+/// Needs to be present for shuttle to be elimination hijacked
+#define ELIMINATION_ENABLED 1
+/// Prevents elimination hijack same way as non-antags
+#define ELIMINATION_PREVENT 2
+
```
Comment (park66665) [inline code/__DEFINES/antagonists.dm:57]: There are HIJACK_NEUTRAL and other defines around 20 lines above. As this PR replaces those defines with these defines, you should delete them.

https://github.com/BeeStation/BeeStation-Hornet/blob/f6059b7d06b3380885a264e8d4842b7d58ef2fac/code/__DEFINES/antagonists.dm#L35-L38

File: code/modules/antagonists/ert/ert.dm [hunk 2]
```diff
@@ -11,10 +11,13 @@
 	var/role = "Security Officer"
 	var/list/name_source
 	var/random_names = TRUE
+	var/rip_and_tear = FALSE
+	var/equip_ert = TRUE
+	var/forge_objectives_for_ert = TRUE
```
Comment (park66665) [inline code/modules/antagonists/ert/ert.dm]: These three variables currently have no use.

**Review by park66665 — COMMENTED** [review]
I only checked if the porting was done properly. Things were mostly okay in that respect.

**Review by utilbird — DISMISSED** [review]
Your branch name hurts my soul :(

**Comment by utilbird** [comment]
@PowerfulBacon needs a tick from you

**Comment by utilbird** [comment]
BACON

File: code/__HELPERS/text.dm [hunk 3]
```diff
@@ -825,3 +825,18 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
...
+	var/static/list/whitespace = list(" ", "\n", "\t")
+	for(var/i in 1 to length(original))
+		var/char = original[i]
+		if(!replace_whitespace && (char in whitespace))
+			out += char
+			continue
+		if(replace_letters_only && (!ISINRANGE(char, 65, 90) && !ISINRANGE(char, 97, 122)))
+			out += char
+			continue
+		out += prob(replaceprob)? pick(replacementchars) : char
```
Comment (utilbird) [inline code/__HELPERS/text.dm]:
```suggestion
		out += prob(replaceprob) ? pick(replacementchars) : char
```

Comment (Sinestia) [reply code/__HELPERS/text.dm]: they call it the tersery operator

**Comment by Sinestia** [comment]
a

**Comment by Sinestia** [comment]
ok got around 2 fixxee

**Comment by utilbird** [comment]
@PowerfulBacon

**Comment by Moja10** [comment]
One fun change would be to make it so that if a shuttle is successfully hijacked, you could make it dock on the nukie dock rather than CC. That way they actually did in fact, hijack the shuttle to a new location

**Comment by Sinestia** [comment]
> 
> 
> One fun change would be to make it so that if a shuttle is successfully hijacked, you could make it dock on the nukie dock rather than CC. That way they actually did in fact, hijack the shuttle to a new location

it does

**Comment by Sinestia** [comment]
Bacon moment

**Review by utilbird — APPROVED** [review]

**Comment by Crossedfall** [comment]
Bacon is absent, so I'm bypassing their code owner requirement per @TheFakeElon 's approved review.

**Comment by Sinestia** [comment]
Thoughts and prayers for uni corgi that he may get the big GPAs

**Pull request merged**